apps/aesophia/test/contracts/erc20_token.aes

```diff
@@ -0,0 +1,97 @@
+contract ERC20Token =
+  record state = {
+    totalSupply : int,
+    decimals : int,
+    name : string,
+    symbol : string,
+    balances : map(address, int),
+    allowed : map(address, map(address,int)),
+    // Logs, remove when native Events are there
+    transfer_log : list((address,address,int)),
+    approval_log : list((address,address,int))}
+
+  // init(100000000, 10, "Token Name", "TKN")
+  public stateful function init(_totalSupply : int, _decimals : int, _name : string, _symbol : string ) = {
+    totalSupply = _totalSupply,
+    decimals = _decimals,
+    name = _name,
+    symbol = _symbol,
+    balances = {[Call.caller] = _totalSupply }, // creator gets all Tokens
+    allowed = {},
+    // Logs, remove when native Events are there
+    transfer_log = [],
+    approval_log = []}
+
+  public stateful function totalSupply() : int = state.totalSupply
+  public stateful function decimals() : int = state.decimals
+  public stateful function name() : string = state.name
+  public stateful function symbol() : string = state.symbol
+
+  public stateful function balanceOf(tokenOwner : address ) : int = lookupDefault(tokenOwner, state.balances, 0)
+
+  public stateful function transfer(to : address, tokens : int) =
+    put( state{balances[Call.caller] = sub(state.balances[Call.caller], tokens) })
+    put( state{balances[to] = add(lookupDefault(to, state.balances, 0), tokens) })
+    transferEvent(Call.caller, to, tokens)
+    true
+
+  public stateful function approve(spender : address, tokens : int) =
+    // allowed[Call.caller] field must have a value!
+    ensure_allowed(Call.caller)
+    put( state{allowed[Call.caller][spender] = tokens} )
+    approvalEvent(Call.caller, spender, tokens)
+    true
+
+  public stateful function transferFrom(from : address, to : address, tokens : int) =
+    put( state{ balances[from] = sub(state.balances[from], tokens) })
+    put( state{ allowed[from][Call.caller] = sub(state.allowed[from][Call.caller], tokens) })
```

Comment:
I see this function mixes checks and effects. I.e. balance subtracted already though allowance checked now.

If execution stops in between because out of gas, I see that [state trees are kept to the initial version](https://github.com/aeternity/epoch/blob/85f5050fbdf4369385ba33a8b51ff8dfed044017/apps/aecontract/src/aect_dispatch.erl#L163) so no impact of partial effect. Ok. This contract in `sub` definition exhausts gas (`require/abort` compiled to `switch(0) 1 => ()`) if check fails. Ok.

Though if execution stopped in between because of REVERT then [the state trees would be not kept to the initial version](https://github.com/aeternity/epoch/blob/85f5050fbdf4369385ba33a8b51ff8dfed044017/apps/aecontract/src/aect_dispatch.erl#L155) (I [tracked this REVERT behaviour](https://www.pivotaltracker.com/story/show/160721302)). When [`abort` is implemented in Sophia as REVERT](https://www.pivotaltracker.com/story/show/159522821) [probably this contract will be updated](https://www.pivotaltracker.com/story/show/159522821/comments/194537013) potentially exposing issue that from gets tokens removed (destroyed?) by user not allowed.

Reply:
Tracked with https://www.pivotaltracker.com/story/show/160751003 - for when there is Sophia abort()